### PR TITLE
net: WIP: Dynamic IPs

### PR DESCRIPTION
**Pull request by Tetrix42**
Currently the Wallet does not update its external IP, my ISP gives me dynamic IP addresses, so my wallet will not run as full node.

what is does:
- if external connections report a different ip than the known external ip, a new discovery is triggered.
- closed connections reduce the score for that IP.
- To increas the score of new IP addresses, the selection uses  randomness.

### CODE REVIEW

File: src/net.cpp
Comment (cyrossignol) [inline src/net.cpp]: There's a security issue here. `CNode::PushVersion()` is called when another node connects. A network of malicious nodes could connect/disconnect quickly to force the victim to query the external IP address lookup service used by `Discover()` and/or the UPnP router repeatedly in quick succession by spoofing the local address in the message. Under enough pressure, this could exhaust the victim's system resources or trigger a blacklisting from the external services. 

Please take a look at [commit `845c86d1` in Bitcoin](https://github.com/bitcoin/bitcoin/pull/5161/commits/845c86d128fb97d55d125e63653def38729bd2ed) that uses a different strategy to detect and re-advertise external IP address changes. It also removes the problematic dependency on the external service. We can probably adopt some of these ideas until we have time in the future to port all of the newer net processing code.

File: src/net.cpp
Comment (cyrossignol) [inline src/net.cpp]:
```suggestion
                        LOCK(cs_mapLocalHost);
                        auto iter = mapLocalHost.find(pnode->addrLocal);
                        
                        if (iter != mapLocalHost.end() && iter->second.nScore > 1)
                        {
                            --iter->second.nScore;
                        }
```

The value of `pnode->addrLocal` is untrusted input supplied by the peer. When we use the subscript operator (`[]`) on `std::map` objects, the map will create a new record if the key doesn't exist. This enables malicious nodes to potentially fill up `mapLocalHost` with bogus addresses. Instead, we can use `std::map::find()` to check for existence first. 

We also need to lock `cs_mapLocalHost` because other threads access and modify the map.

Comment (cyrossignol) [reply src/net.cpp]: Speaking of bogus entries, should we remove `pnode->addrLocal` from `mapLocalHost` when it's no longer valid? If the node's external IP address changes often, iteration over `mapLocalHost` in `GetLocal()` will gradually slow down as stale addresses build up. Does `nScore` ever drop below zero?

**Review by cyrossignol — COMMENTED** [review]
@Tetrix42 thanks for working on this! I noticed this problem before as well.

There are a few moving parts to the local address selection for each peer. I need to do a closer review and some testing to see how your new algorithm in `GetLocal()` interacts with these, but it looks fine so far. 

I added some initial comments about security concerns for the other bits. Please let me know what you think.

File: src/net.cpp
Comment (jamescowens) [reply src/net.cpp]: I think we should remove stale addresses from the map.

File: src/net.cpp
Comment (Tetrix42) [reply src/net.cpp]: dropped `Discover()`, so the problem with blacklisting is solved.
The addresses reported are now stored in a map like `map<IP_remote, IP_reportet>`.

If Mallory pushes tells us different addresses only the most recent one is stored, thus casts only one vote on our address.
(unless Mallory also changes its address)

So we will rely on the answers of our peers.

File: src/net.cpp
Comment (Tetrix42) [reply src/net.cpp]: with the new map each peer has one vote, that is stored in the map, and on disconnect removed, alse each peer will only create one entry and not more.
I think i read something of connection limit of 250, or somewhere in that range, so the map will hold up to that ammount, i think that is acceptable now.

Comment (jamescowens) [reply src/net.cpp]: Actually, we have addnodes that are supporting 750+ connections, but I think that is still ok.

**Comment by Tetrix42** [comment]
ping. last update was a week ago, my client in testnet behaves as i expect.

**Comment by cyrossignol** [comment]
Sorry Tetrix, I haven't forgotten about this. Just did not have a chance 
to dig into the updates yet--been really busy the last few days. I'll 
try to take a look tonight or tomorrow.

**Review by jamescowens — CHANGES_REQUESTED** [review]
@Tetrix42 This is a good effort. @cyrossignol and I really would like you to more closely follow Bitcoin on this. The reason is that Bitcoin has had a protocol expert working on the net code and refining it. We are slowly upgrading our core code to Bitcoin's current standards, and we need to ensure when we move the ball downfield in the core areas, such as the P2P networking, we follow them as precisely as we can.

I like this PR and would like you to continue working on it, getting the implementation as close to Bitcoin's as possible. I recognize some things will have to be different due to the older net code we have, but this area is fairly well encapsulated, so should backport well.

I think we will shoot for merging this PR after Fern (i.e. Gladys), because I am nervous about messing with this code with all of the other changes we are doing.

I would like you specifically to run a testnet node with the most up date version of this PR rebased on top of development as we finish out Fern, and you can report on its stability as we get through Fern and start working on Gladys. That will allow us to get a head start on determining stability and the interaction with all of the new code in Fern.

**Comment by Tetrix42** [comment]
Thanks for the reply.
OK i will modify it to be closer to bitcoins code.
might take some time.

**Comment by jamescowens** [comment]
No problem. Glad you worked on this. You have some time. We have to get Fern out of the door. The good thing is that this part of the code should not change in the remaining work for Fern, so rebasing should be a snap.

**Review by cyrossignol — COMMENTED** [review]
Looks great so far--I think you're pretty close. 

This probably won't react as quickly to IP address changes as your original implementation. Like @jamescowens said, though, this helps tremendously to move us closer to parity with Bitcoin so that we can eventually do a full port of the newer code base (it was a silly decision to fork Gridcoin without maintaining the parent repository relationship), and this is an important update.

I'll wait for your signal before testing and leaving comments.

File: src/main.cpp
Comment (Tetrix42) [inline src/main.cpp:3914]: If `AddLocal` in net.cpp was successfull e.g. by usage with `-upnp` then GetLocalAddress will always return this address. any code that calls AddLocal is only called during startup, thus LocalAddresses are not updated.
If only local addresses are discovered, e.g., without `-upnp` and behind NAT of the router, then no addresses are stored, and the check above will fail. Then the check of the following `else if` will pass, and an external IP is guessed from the other node.

Dynamic of IPs is not part of Bitcoin either.
The discovery Threads (UPNP and interface IP) has to be updated, that addresses are checked every 20 minutes or so, as UPNP uses this interval for portmapping anyway. Deletion of old IPs also needs to be added.

File: src/net.cpp
Comment (Tetrix42) [inline src/net.cpp:1104]: to be removed

File: src/net.cpp
Comment (Tetrix42) [inline src/net.cpp:458]: addresses seen by peers are not added, but only increase the confidence in an address.
Addresses need to be added by other sources e.g. UPNP or from the Interfaces.
But if there are no addresses added, because UPNP is not used and no external addresses are found on the Intefraces, the address the peer told us is sent directly back to the peer, and no voting happens, thats somewhat inconsequent.

File: src/main.cpp
Comment (Tetrix42) [inline src/main.cpp:4782]: In `AdvertiseLocal` is currently the only place a new IP could be chosen to be pushed, but only every 24, and only by chance.
My ISP forces me to use a new IP after 24 hours, so the interval should be reduced, but if used on servers with static IP it might be spam. what are your opinions?
Currently there is also a 2h interval that forces disconnects in `net.cpp:1203` that is marked as `consider for removal` but i think it could help to keep a dynamic in the network and prevent it to split.
So also the 2h interval is an argument to reduce this advertisement interval.

**Review by Tetrix42 — COMMENTED** [review]
After testing the code, and looking at the source of Bitcoin (did not set-up a bitcoin node), it looks like they did not really solve the dynamicIP issue, but only removed external IP providers. waiting for up to 24 hours to send the new IP while the IP is only valid for 24 hours is not so useful. or the voting for addressse is not executed correcty. (see one of the next comments)

i also noticed strange/unexpected/unwanted behaviour during testing with tor (but it might be out of scope for this MR)

there are some points we need to discuss on.

to archive solve the dynamicIP problem i liked my previous attempt more.

File: src/main.cpp
Comment (cyrossignol) [reply src/main.cpp:4782]: @Tetrix42 Thanks for your analysis. Your comments are correct. Bitcoin's approach is passive and mostly depends on a node's outbound connections to propagate dynamic IP address reassignments. When your node connects to another listening node, that peer records your current IP address and then&mdash;through peer exchange&mdash;sends it to other nodes. Of course, that only works when your node pushes back the IP address that the peer reported in `addrLocal`&mdash;it won't work very well when your node sends the IP address assigned by UPnP after that IP address changes.

As a result, the current "visible" IP address of a listening node with dynamic reassignment is often different from the inventory of IP addresses that the node detected for itself during startup via `AddLocal()`. This interaction is less reactive to IP address changes than the direct approach that you originally proposed, but it reduces overhead, mutable state, and attack surface by relying on the existing peer discovery protocol. 

Personally, I think it's a rather messy way to solve the problem. We can improve this area after we merge Bitcoin's newer peer-to-peer networking code, including some of your original ideas. Since most listening nodes bind to IP addresses that change very infrequently, I'm not sure that it's worth introducing extra overhead to handle this at the moment. Perhaps we can make the interval for `AdvertiseLocal()` configurable (with a default at 24 hours) for nodes that need to push updates more frequently. 

> Currently there is also a 2h interval that forces disconnects in net.cpp:1203 that is marked as consider for removal but i think it could help to keep a dynamic in the network and prevent it to split.

I agree&mdash;the connection cycling is a nice property for decentralization. However, it also makes it easier for an actor to gradually take control of inbound connections on listening nodes to partition the network or execute a DoS attack. Bitcoin has more advanced routines now for connection eviction that attempt to balance these concerns...another reason to update our code base.

**Comment by jamescowens** [comment]
@Tetrix42, we haven't forgotten this. We are just busy getting Fern out the door. When we get that released, lets come back to this, get this rebased, which shouldn't be all that hard, and see where we are.

**Comment by jamescowens** [comment]
@Tetrix42, can you rebase this now? Let's pick this back up again now that Fern is out the door! :)

**Comment by jamescowens** [comment]
@cyrossignol we should also discuss whether we should merge this as a "lead-in" to the adoption of the newer Bitcoin net code. That seems to be the best approach.

**Comment by jamescowens** [comment]
@Tetrix42 are you around?

**Comment by Tetrix42** [comment]
> @Tetrix42 are you around?
sorry, I was busy latley, i should have time to rebase and include bitcoins new changes until end of next weekend.

**Comment by Tetrix42** [comment]
Sorry for doing a new pull request (https://github.com/gridcoin-community/Gridcoin-Research/pull/2002), but with the progress in the Gridcoin source it was easier to implement the wishes you had regarding the bitcoin net.cpp in a new pull request, directly based on the current development.

**Pull request closed without merging**
